lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -580,92 +690,648 @@ void crypto_bignum_clear(struct bignum *s)
+	/* Reset mpi to skip freeing here, those mpis will be freed with key */
+	mbedtls_mpi_init(&rsa.E);
+	mbedtls_mpi_init(&rsa.N);
+	mbedtls_rsa_free(&rsa);
+
+	return res;
 }
 
-TEE_Result crypto_acipher_rsanopad_encrypt(struct rsa_public_key *key __unused,
-					   const uint8_t *src __unused,
-					   size_t src_len __unused,
-					   uint8_t *dst __unused,
-					   size_t *dst_len __unused)
+TEE_Result crypto_acipher_rsanopad_decrypt(struct rsa_keypair *key,
+					   const uint8_t *src, size_t src_len,
+					   uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	mbedtls_rsa_context rsa;
+	int lmd_res;
+	uint8_t *buf = NULL;
+	unsigned long blen, offset;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+
+	rsa.E = *(mbedtls_mpi *)key->e;
+	rsa.D = *(mbedtls_mpi *)key->d;
+	rsa.N = *(mbedtls_mpi *)key->n;
+	if (key->p && crypto_bignum_num_bytes(key->p)) {
+		rsa.P = *(mbedtls_mpi *)key->p;
+		rsa.Q = *(mbedtls_mpi *)key->q;
+		rsa.QP = *(mbedtls_mpi *)key->qp;
+		rsa.DP = *(mbedtls_mpi *)key->dp;
+		rsa.DQ = *(mbedtls_mpi *)key->dq;
+	}
+
+	rsa.len = mbedtls_mpi_size(&rsa.N);
+
+	blen = LMD_MAX_BITS_PER_VARIABLE / 8;
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	memset(buf, 0, blen);
+	memcpy(buf + rsa.len - src_len, src, src_len);
+
+	lmd_res = mbedtls_rsa_private(&rsa, NULL, NULL, buf, buf);
+	if (lmd_res != 0) {
+		FMSG("mbedtls_rsa_private() returned 0x%x", -lmd_res);
+		res = get_tee_result(lmd_res);
+		goto out;
+	}
+
+	/* Remove the zero-padding (leave one zero if buff is all zeroes) */
+	offset = 0;
+	while ((offset < rsa.len - 1) && (buf[offset] == 0))
+		offset++;
+
+	*dst_len = rsa.len - offset;
+	memcpy(dst, (char *)buf + offset, *dst_len);
+out:
+	if (buf)
+		free(buf);
+	/* Reset mpi to skip freeing here, those mpis will be freed with key */
+	mbedtls_mpi_init(&rsa.E);
+	mbedtls_mpi_init(&rsa.D);
+	mbedtls_mpi_init(&rsa.N);
+	if (key->p && crypto_bignum_num_bytes(key->p)) {
+		mbedtls_mpi_init(&rsa.P);
+		mbedtls_mpi_init(&rsa.Q);
+		mbedtls_mpi_init(&rsa.QP);
+		mbedtls_mpi_init(&rsa.DP);
+		mbedtls_mpi_init(&rsa.DQ);
+	}
+	mbedtls_rsa_free(&rsa);
+
+	return res;
 }
 
-TEE_Result crypto_acipher_rsaes_decrypt(uint32_t algo __unused,
-					struct rsa_keypair *key __unused,
+TEE_Result crypto_acipher_rsaes_decrypt(uint32_t algo, struct rsa_keypair *key,
 					const uint8_t *label __unused,
 					size_t label_len __unused,
-					const uint8_t *src __unused,
-					size_t src_len __unused,
-					uint8_t *dst __unused,
-					size_t *dst_len __unused)
+					const uint8_t *src, size_t src_len,
+					uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int lmd_res, lmd_padding;
+	size_t blen;
+	size_t mod_size;
+	void *buf = NULL;
+	mbedtls_rsa_context rsa;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo = MBEDTLS_MD_NONE;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
```

Comment:
This line and the ~10 lines below is a recurring pattern, how about adding a helper function:
```C
static void rsa_init_from_key_pair(mbedtls_rsa_context *rsa, struct rsa_keypair *key);
```

Reply:
Good suggestion, but I will add one more parameter 'type' which indicate encrypt or decrypt operate.